Improve account reference

lomamech opened the pull request:
- Remove of the property debit_balance and credit_balance on
the reference of Account
- Give the possibility to be able to delete the parent property in the
  registration form of the reference accounts
- Implementing the Account Reference Type module
- Implementing the server API
- Implementation of the integration test as well as the E2E test for the
management of the account reference type
- Complete Integration test and E2E test for Account Reference
- Implement Component Account Reference type Select
- Set Default value for Account Reference Type
- Add the component for the select of Account Reference Type
- Update the report of Account Reference Set the Account Reference type if
  is selected

closes #3523

lomamech committed:
refactor(Reference Account)
- Remove of the property debit_balance and credit_balance on
the reference of Account

lomamech committed:
improvement(Account Reference)
- Give the possibility to be able to delete the parent property in the
  registration form of the reference accounts

lomamech committed:
improvement(Account Reference)
- Complete Integration test and E2E test for Account Reference
- Implement Component Account Reference type Select
- Set Default value for Account Reference Type

lomamech committed:
Sanitaze Code

lomamech committed:
refactor(Account Reference Report)
- Add the component for the select of Account Reference Type
- Update the report of Account Reference Set the Account Reference type if
  is selected

lomamech committed:
feature(Migrate Database file)
- Set Query for the migration of Database Schema and set default Date

lomamech committed:
Sanitaze Code

lomamech committed:
Complete migration file

jniles commented on client/src/modules/account_reference/account_reference.modal.js:
```diff
@@ -52,14 +52,22 @@ function AccountReferenceModalController($state, Accounts, AccountReferences, No
   // submit the data to the server from all two forms (update, create)
   function submit(accountReferenceForm) {
     if (accountReferenceForm.$invalid) { return null; }
+    // Fixe me @lomamech : Give the possibility to validate the modification
```
:+1:

jniles commented on client/src/modules/account_reference_type/account_reference_type.js:
```diff
@@ -0,0 +1,103 @@
+    AccountReferenceType.read()
+      .then((accountReferenceType) => {
+        accountReferenceType.forEach((item) => {
+          item.label = $translate.instant(item.label);
```
It looks like every time we download account reference types, we want to translate them.  Let's move this operation into the `AccountReferenceTypeService` so that after every `read()` call, we translate these labels.

jniles commented on client/src/modules/account_reference_type/modals/account_reference_type.modal.html:
```diff
@@ -0,0 +1,47 @@
+    <!-- label -->
+    <div class="form-group" ng-class="{ 'has-error' : AccountReferenceTypeForm.$submitted && AccountReferenceTypeForm.description.$invalid }">
+      <label class="control-label" translate>FORM.LABELS.DESIGNATION</label>
+      <div>
```
Why do we need this extra div..?

jniles commented on client/src/modules/account_reference_type/modals/account_reference_type.modal.html:
```diff
@@ -0,0 +1,47 @@
+    <ol class="headercrumb">
+      <li class="static" translate>TREE.ADMIN</li>
+
+      <li ng-if="AccountReferenceTypeModalCtrl.isCreating" class="title">
```
It looks like the only thing that changes here is the `badge`.  Could you move the `ngIf` to put it on the badge?

```html
<label ng-if="AccountReferenceTypeModalCtrl.isCreating" class="badge badge-warning" translate>FORM.LABELS.CREATE</label>
```

jniles commented on client/src/modules/account_reference_type/modals/account_reference_type.modal.html:
```diff
@@ -0,0 +1,47 @@
+    </ol>
+  </div>
+  
+  <div class="modal-body" style="overflow-y: scroll; max-height:600px; ">
```
NOTE: We should try to standardize our modal sizes somewhere.

jniles commented on client/src/modules/account_reference_type/templates/action.cell.html:
```diff
@@ -0,0 +1,36 @@
+    </a>
+
+    <ul data-row-menu="{{row.entity.label}}" class="dropdown-menu-right" bh-dropdown-menu-auto-dropup uib-dropdown-menu>	
+      <li>
```
Could you add a header to the dropdown to give the user context about which row they have clicked?  It should look like [the one found on the account action dropdown](https://github.com/IMA-WorldHealth/bhima/blob/master/client/src/modules/accounts/templates/grid.actionsCell.tmpl.html#L9).

jniles commented on client/src/modules/account_reference_type/templates/action.cell.html:
```diff
@@ -0,0 +1,36 @@
+    <ul data-row-menu="{{row.entity.label}}" class="dropdown-menu-right" bh-dropdown-menu-auto-dropup uib-dropdown-menu>	
+      <li>
+        <a data-method="edit-record" ng-click="grid.appScope.edit(row.entity)" href>
+          <span>
```
Do we need this extra `<span>`?

jniles commented on server/controllers/finance/accounts/accountReferenceType.js:
```diff
@@ -0,0 +1,109 @@
+
+// PUT /account_reference_type /:id
+function update(req, res, next) {
+  const sql = `UPDATE account_reference_type SET ? WHERE id = ?;`;
```
Let's delete `req.body.id` before doing this action to make sure the user cannot change the id.

jniles commented on server/controllers/finance/accounts/accountReferenceType.js:
```diff
@@ -0,0 +1,109 @@
+}
+
+// DELETE /account_reference_type/:id
+function del(req, res, next) {
```
Let's cal this function `remove` instead of `del` since `delete` is a keyword.

jniles reviewed (commented):
Just a few comments here.  Looks pretty good!

lomamech committed:
refactor(Account Reference Type)
- Let move translation in Account Reference Type Service
- Sanitaze Code
- Add header to the Dropdown

mbayopanda commented on client/src/js/components/bhAccountReferenceTypeSelect.js:
```diff
@@ -0,0 +1,33 @@
+  };
+
+  $ctrl.onSelect = accountReferenceType => $ctrl.onSelectCallback({ accountReferenceType });
+}
```
👍

mbayopanda commented on client/src/modules/account_reference_type/account_reference_type.js:
```diff
@@ -0,0 +1,99 @@
+  }
+
+  function loadGrid() {
+    toggleLoadingIndicator();
```
We don't know the value of the `vm.loading` here, it will be more clear to assign the value directly to the variable instead of calling the function. For example : 
```js
vm.loading = true;
```

mbayopanda commented on client/src/modules/account_reference_type/modals/account_reference_type.modal.html:
```diff
@@ -0,0 +1,42 @@
+  </div>
+
+  <div class="modal-footer">
+    <p id="account-reference-type-same" ng-if="AccountReferenceTypeForm.$submitted && AccountReferenceTypeForm.$pristine && !AccountReferenceTypeModalCtrl.isCreating" class="text-warning">
```
Is it possible to have a variable for example `vm.hasNoChange` in the controller like this : 
```js
vm.hasNoChange = form.$submitted && form.$pristine && !vm.isCreating;
```
So that we can reduce the code in the view like this : 
```html
<p id="account-reference-type-same" ng-if="AccountReferenceTypeModalCtrl.hasNoChange" class="text-warning">
  <i class="fa fa-warning"></i> <span translate>FORM.INFO.RECORD_SAME</span>
</p>
```

mbayopanda replied:
Also, the translation entry key 'FORM.INFO.RECORD_SAME' can have a meaningful name which reflect the message : 
![image](https://user-images.githubusercontent.com/5445251/51823400-925b0200-22df-11e9-8f86-9c40ecb4298b.png)

Something like : `FORM.INFO.NO_CHANGE` can be not bad.

mbayopanda commented on client/src/modules/account_reference_type/modals/account_reference_type.modal.js:
```diff
@@ -0,0 +1,72 @@
+  // submit the data to the server from all two forms (update, create)
+  function submit(accountReferenceTypeForm) {
+    if (accountReferenceTypeForm.$invalid) { return null; }
+    if (!accountReferenceTypeForm.$dirty) { return null; }
```
lines above can be written like this : 
```js 
if (accountReferenceTypeForm.$invalid || !accountReferenceTypeForm.$dirty) { return null; }
```

mbayopanda commented on client/src/modules/reports/generate/account_reference/account_reference.html:
```diff
@@ -34,6 +34,14 @@ <h3 class="text-capitalize" translate>REPORT.ACCOUNT_REFERENCE.TITLE</h3>
+              id="reference_type_id"
+              name="reference_type_id"
+              on-select-callback="ReportConfigCtrl.onSelectAccountReferenceType(accountReferenceType)">
+              <bh-clear on-clear="ReportConfigCtrl.clear('reference_type_id')"></bh-clear>
```
👍 well done

mbayopanda commented on client/src/modules/templates/bhAccountReferenceTypeSelect.tmpl.html:
```diff
@@ -0,0 +1,27 @@
+    <ui-select
+      name="reference_type_id"
+      ng-model="$ctrl.referenceTypeId"
+      on-select="$ctrl.onSelect($item, $model)">
```
Could you add the property `append-to-body="true"` so that the ui-select list be correctly displayed in the modal.

![image](https://user-images.githubusercontent.com/5445251/51825279-b66d1200-22e4-11e9-9918-aad81a5fb06f.png)

mbayopanda commented on server/controllers/finance/accounts/accountReferenceType.js:
```diff
@@ -0,0 +1,113 @@
+    .then((row) => {
+    // if nothing happened, let the client know via a 404 error
+      if (row.affectedRows === 0) {
+        throw new NotFound(`Could not find a function with id ${req.params.id}`);
```
Does it correct the error message talk about function ?

mbayopanda replied:
It is about preference may be, did you try the `lookupAccountReferenceType(req.params.id)` function for looking if the account reference exists, this function will thrown an 404 error automatically if the record doesn't exists.

If you decide to use the `lookupAccountReferenceType` function the code can be : 
```js
const sql = `DELETE FROM account_reference_type WHERE id = ?;`;
lookupAccountReferenceType(req.params.id)
  .then(() => db.exec(sql, [req.params.id]))
  .then(() => res.sendStatus(204))
  .catch(next)
  .done();

```

mbayopanda commented on server/controllers/finance/reports/account_reference/report.handlebars:
```diff
@@ -16,6 +16,12 @@
 
+      <h4 class="text-center text-uppercase">
+        <strong>
+          {{translate referenceTypeLabel}}
```
👍

mbayopanda commented on server/models/migrations/v0.8.0-v0.9.0/migrate.sql:
```diff
@@ -1021,3 +1021,26 @@ INSERT INTO report VALUES  (23, 'unpaid-invoice-payments', 'REPORT.UNPAID_INVOIC
+INSERT INTO `account_reference_type` (`id`, `label`, `fixed`) VALUES
+(1, 'FORM.LABELS.FEE_CENTER', 1),
+(2, 'FORM.LABELS.BALANCE_SHEET', 1),
+(3, 'FORM.LABELS.PROFIT_LOSS', 1);
```
It is a good idea to have translation key as label, but the bad thing with that is if `FORM.LABELS.BALANCE_SHEET` is translated  `Bilan` we can create a new account reference type with the label `Bilan` the user will see duplicated data : 

![image](https://user-images.githubusercontent.com/5445251/51828067-0e0e7c00-22eb-11e9-9cc4-b0b3004034aa.png)

Even if in the database we have : 
![image](https://user-images.githubusercontent.com/5445251/51828112-2bdbe100-22eb-11e9-9d15-eb85df6c4a7b.png)

We can just set fixed account reference type with real label not with translation key so that we have uniform value in the database (not translation key mixed with real account reference type name).

```sql
INSERT INTO `account_reference_type` (`id`, `label`, `fixed`) VALUES
(1, 'Centre de cout', 1),
(2, 'Bilan', 1),
(3, 'Centre de profit', 1);
```

lomamech replied:
If I set the default values in French, what will happen if we install Bhima in an English speaking country? I kept the same logic that exists on the transaction type table

mbayopanda reviewed (commented):
The code LGTM 👍  

I made some comments for clarity and better user experience.

lomamech committed:
improvement(Account Reference)
- Improve, Revue and sanitaze code after review

lomamech committed:
Resolve Conflict

lomamech committed:
test(Account Reference)
- Update and Improve component for selected bhAccountReferenceType in E2E
  test

lomamech committed:
improvement(Account Reference Type)
- assign the value of vm.loading true and set it false after loading

lomamech committed:
improve(Account Reference)
- Sanitaze code, remove not necessary function toggleLoadingIndicator
